lib/orocos/process.rb

```diff
@@ -1210,6 +1253,21 @@ def self.wait_running(process, timeout = nil, name_service = Orocos::CORBA.name_
             end
         end
 
+        def parse_ior_message(message)
+            if message.include? "\n"
+                message_no_line_end = message.split(/\n/).first
+                if message_no_line_end.include? " "
+                    task_name, ior = message_no_line_end.split(/ /)
+                    return task_name, ior, :FULL_IOR
+                end
+
+                if message_no_line_end == "DONE"
+                    return nil, nil, :DONE
+                end
+            end
+            return nil, nil, :PARTIAL
+        end
+
```

Comment:
I thought we said we would go for JSON to avoid having to do this ?

Plus, really, you want to either have all or none. No need to deal with partials. Even more so as the orogen-generated code writes the whole string at the end... I.e. all the tasks have already been initialized.

- read all data that is in the pipe using read_nonblock(1024) in a loop. Careful, the method raises IO::WaitReadable when the read would block. Put all that data in a single string.
- if ior_read.eof? check whether the data is valid (i.e. valid JSON and contains all expected tasks)
  - if it is valid, set the hash internally and let all the methods resolve tasks
  - if it is not valid, raise that there is something wrong
 - if !ior_read.eof? wait until the next call to wait_running

Reply:
> I thought we said we would go for JSON to avoid having to do this ?

I...forgot about that.

> read all data that is in the pipe using read_nonblock(1024) in a loop. Careful, the method raises IO::WaitReadable when the read would block. Put all that data in a single string.

Ah okay, after reading ruby docs I wasnt sure if read_nonblock 1024 would block or not. Thanks for clearing out how it works.

Reply:
> read all data that is in the pipe using read_nonblock(1024) in a loop. Careful, the method raises IO::WaitReadable when the read would block. Put all that data in a single string.

Should I really do it in a loop inside of `wait_running`? It is already called inside a loop, and I could risk messing with the method's timeout by doing this.

Reply:
Yes, because we want to optimize the zero-timeout case, which is called in a "slow" loop (the Syskit event loop).

It won't mess with the timeout as you are calling nonblock, and because the writer has a finite amount of data to write.

Reply:
> if read_nonblock 1024 would block or not.

It's good to be careful ... but assuming that a method that is called `_nonblock` would block is a bit paranoid.